qa/rpc-tests/wallet_listreceived.py

```diff
@@ -380,9 +394,71 @@ def test_received_sapling(self, height):
         c = self.nodes[1].z_getnotescount(0)
         assert_equal(3, c['sapling'], "Count of unconfirmed notes should be 3(2 in zaddr1 + 1 in zaddr2)")
 
+        # As part of UA support, a transparent address is now accepted
+        r = self.nodes[0].z_listreceivedbyaddress(taddr, 0)
+        assert_equal(len(r), 1)
+        assert_equal(r[0]['pool'], 'transparent')
+        assert_equal(r[0]['txid'], txid_taddr)
+        assert_equal(r[0]['amount'], Decimal('4'))
+        assert_equal(r[0]['amountZat'], 400000000)
+        assert_equal(r[0]['confirmations'], 3)
+        assert r[0]['outindex'] < 2
+        
+        # Test unified address
+        node = self.nodes[1]
+
+        # Create a unified address on one node, try z_listreceivedbyaddress on another node
+        account = self.nodes[0].z_getnewaccount()['account']
+        r = self.nodes[0].z_getaddressforaccount(account)
+        unified_addr = r['unifiedaddress']
+        # this address isn't in node1's wallet
+        assert_raises_message(
+            JSONRPCException, 
+            "From address does not belong to this node",
+            node.z_listreceivedbyaddress, unified_addr, 0)
+
+        # create a UA on node1
+        r = node.z_getnewaccount()
+        account = r['account']
+        r = node.z_getaddressforaccount(account)
+        unified_addr = r['unifiedaddress']
+        receivers = node.z_listunifiedreceivers(unified_addr)
+        assert_equal(len(receivers), 2)
+        assert 'transparent' in receivers
+        assert 'sapling' in receivers
+        # Wallet contains no notes
+        r = node.z_listreceivedbyaddress(unified_addr, 0)
+        assert_equal(len(r), 0, "unified_addr should have received zero notes")
+
+        # Create a note in this UA on node1
+        opid = node.z_sendmany(zaddr1, [{'address': unified_addr, 'amount': 0.1}])
+        txid_sapling = wait_and_assert_operationid_status(node, opid)
+        self.generate_and_sync(height+5)
+
+        # Create a UTXO that unified_address's transparent component references, on node1
+        outputs = {receivers['transparent']: 0.2}
+        txid_taddr = node.sendmany("", outputs)
+
+        r = node.z_listreceivedbyaddress(unified_addr, 0)
+        assert_equal(len(r), 2, "unified_addr should have received 2 payments")
+        assert_equal(r[0]['pool'], 'sapling')
+        assert_equal(r[0]['txid'], txid_sapling)
+        assert_equal(r[0]['amount'], Decimal('0.1'))
+        assert_equal(r[0]['amountZat'], 10000000)
+        assert_false(r[0]['change'], "Note valued at 0.1 should not be change")
+        assert_equal(r[0]['memo'], no_memo)
+
+        assert_equal(r[1]['pool'], 'transparent')
+        assert_equal(r[1]['txid'], txid_taddr)
+        assert_equal(r[1]['amount'], Decimal('0.2'))
+        assert_equal(r[1]['amountZat'], 20000000)
+        assert_equal(r[1]['confirmations'], 0)
+        assert_equal(r[1]['outindex'], 0)
```

Comment:
Is the order of these outputs in `r` guaranteed? Otherwise this could cause nondeterministic failures.